Fan plots

billetd opened the pull request:
# New Feature 
**Fan plots**

**module:** *fan.py*

**package:** *pydarn/plotting*

## Scope
This PR adds basic functionality for making fan plots from read in fitACF data. 

For a given scan number (dictated by the scan flags within the data list of dictionaries), either LOS velocity, power, spectral width or elevation can overlain on an AACGM latitude - magnetic local time plot. This code utilises the AACGMv2 library by @aburrell, making it now a requirement for pyDARN.

~~Currently, it does not allow for multiple different fans to be overlain on the same plot (either from the same radar or other)~~. It also does not currently support plotting the nearest scan to a given time (only scan numbers are used - i.e. the nth scan since the beginning of records). ~~Both of these features~~ this will be added in a future PR.

You plot your choice of range gates only by setting ranges=[two element list] of the lower and upper boundary. Default is ranges=[0,75].

Fan plotting also currently does not work for non standard beam-gate layouts (e.g. 15km range gates on Borealis radars) or special control programs (e.g. camping beams). It will always assume that the radar data given uses the standard number of beams and gates for that particular radar. For example, if you try to give it some Hankasalmi data with only 2 beams in the data strucutre (for whatever reason), it will plot something weird as it assumes it will be getting 16 beams for that radar. 

## Test code
```python
import matplotlib.pyplot as plt
import pydarn

#Read in fitACF file
file = "20200507.0.fitacf"
SDarn_read = pydarn.SuperDARNRead(file)
fitacf_data = SDarn_read.read_fitacf()

#Make the fan plot and show
fanplot = pydarn.fan.plot_fan(fitacf_data, scan_index=1, groundscatter=1, colorbar='True', parameter='p_l', colorbar_label='Power [dB]')
plt.show()
```

You might notice that the variable `fanplot` in the examples above actually holds some information. This contains the AACGM latitude and longitude of the fan just plotted, as well as the data and ground scatter information. If you instead change `fanplot` to 4 separate variables, it will return the latitude, longitude, data, groundscatter and datetime info into seperate numpy arrays:
```python
lats,lons,data,grndsct,datetime=pydarn.fan.plot_fan(fitacf_data, scan_index=27)

lats.shape

```
Which returns `>>>(76, 17)`, i.e. ranges x beams array of the AACGM latitude, and so on for the other variables. The groundscatter array is 0's and 1's, 1 being a range gate flagged as groundscatter.

Feel free to check out fan.py for more options. E.g., you can change the low latitude boundary to 40 degrees by passing in `lowlat=40'. 

~~As default, fan plots also caps the number of range gates at 75. You can turn this off by setting `all_gates=True`.~~

### Where to obtain test files
Your faveourite SuperDARN data repo

### Edits
2020/06/15
- Problem with picking the actual correct scan chosen fixed
- Fixed issue with manifest file not putting radar_fov_files in correct place when using pip
- Added documentation
- Added automatic detection of the hemisphere to fan.py

2020/06/22
- Renamed `plot_radar_fov` method to `plot_fan`. All references to it should also be fixed.
- Thanks to a suggested change by @aburrell, can now easily plot multiple fans on same plot.
- `all_gates` keyword has been removed, replaced with `ranges`. An inputted two element list of the lower and upper ranges to plot (default is [0,75]).
- Behaviour of colorbar keyword has changed. Now simply "True" or "False" to display colorbar.
- Added the `boundary` keyword. "True" by default, but when set to "False", doesn't plot the outline of the fan plot.
- Updates docs with new features.

2020/08/11
- Fix for radars which are scanning backwards

2020/11/26
- Merged in from develop to include pyDARNio and beam position finding changes.
- Now accepts axis object as a keyword instead of overwriting. If `ax` keyword is left blank, it will make a new polar plot axis object.
- datetime object now also returned

mts299 commented:
Please ignore Travis-CI macOS is breaking on pytest but pydarn installs fine.

kherli commented:
Hei :)

I tested the code for a different fitacf3 file I already had on my system. The data reads okay but I get the following error when I try to call pydarn.fan.plot_radar_fov:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/katie/anaconda3/envs/pydarnFan/lib/python3.7/site-packages/pydarn/plotting/fan.py", line 107, in plot_radar_fov
    beam_corners_lats = np.loadtxt(beam_lats)
  File "/home/katie/anaconda3/envs/pydarnFan/lib/python3.7/site-packages/numpy/lib/npyio.py", line 981, in loadtxt
    fh = np.lib._datasource.open(fname, 'rt', encoding=encoding)
  File "/home/katie/anaconda3/envs/pydarnFan/lib/python3.7/site-packages/numpy/lib/_datasource.py", line 269, in open
    return ds.open(path, mode, encoding=encoding, newline=newline)
  File "/home/katie/anaconda3/envs/pydarnFan/lib/python3.7/site-packages/numpy/lib/_datasource.py", line 623, in open
    raise IOError("%s not found." % path)
OSError: /home/katie/anaconda3/envs/pydarnFan/lib/python3.7/site-packages/pydarn/plotting/../radar_fov_files/090_lats.txt not found.
```

billetd commented:
Thanks for trying it out @kherli. It looks like it probably has to do with these lines (80-81) in `fan.py':

```python
        my_path = os.path.abspath(os.path.dirname(__file__))
        base_path = os.path.join(my_path, '..')
```python

They are "supposed" to find the base install directory of pyDARN so that later it can find the FOV files in the right place, but it looks like in your install it's pointing to the plotting sub-folder. I'll check this out, because it's probably a side-effect of me not knowing Python very well!

billetd commented:
Could you try something for me when you get the chance @kherli? On your install, could you try changing the ``base_line`` variable on line 81 of ``fan.py`` to the base directory of pyDARN, and see if that works? I.e. the folder that contains the ``radar_fov_files`` folder. For some reason I can't reproduce the bug, but I don't use conda.

jpreistad commented:
Great work, I have really looked forward to this plotting functionality! I get the same behavior as @kherli mention above. To me it seems that the radar_fov_files has been placed in a wrong directory by the install process. I have just moved the folder back to the base directory of  pydarn (~/anaconda3/envs/pydarnfan/lib/python3.7/site-packages) which solved the issue for me.

jpreistad commented:
Another thing I noticed is that the code seem to always assume the time of the first beam in the file when calculating MLT (Line 117-120 in fan.py). I will try to make a fix here when I get the time.

kherli commented:
Both @billetd and @jpreistad's suggestions work for me and the fanplot now shows up :)

mts299 commented:
I will do a review of the code. @billetd are you going to do readthedocs documentation after this PR or in this PR?

billetd commented:
@mts299 I'll do the docs as a seperate PR

mts299 commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,248 @@
+        # check out the file. Even has an enum for hemisphere and associated
+        # value you will find in MAP files
+        """
+        Plots a radar FOV fan plot for the given data and scan number
```
``` suggestion
        Plots a radar's Field Of View (FOV) fan plot for the given data and scan number
```

mts299 commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,248 @@
+                Default: v (Velocity). Alternatives: 'p_l', 'w_l', 'elv'
+            hemisphere: str
+                North or south, depending on radar. Currently,
+                pyDARN cannot work this out
```
What do you mean pyDARN cannot work this out? Its in the SuperDARNRadars list.

billetd replied:
Resolved

mts299 commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,248 @@
+            lowlat: int
+                Lower AACGM latitude boundary for the polar plot
+                Default: 50
+            all_gates: bool
```
Could we make this an int and call it `max_gates`? This will allow for more flexibility.

billetd replied:
Potentially, but it would leave the user to guess the maximum number of gates. Could impliment it so that if they choose a number of max gates, it defaults to max gates?

mts299 replied:
yea I think that would be better. I do that with RTP plots.

aburrell replied:
But what about if I only want E-region fans?  How does a user specify a custom number of range gates?  Also, I frequently plot fans for 1/2F ranges only, which means I need to specify both a custom minimum and maximum.  It doesn't look like I can do that here.

billetd replied:
Updated functionality. `all_gates` has been removed and replaced with `ranges`, which is a two element list of the lower and upper ranges to plot. Default is [0,75]

mts299 commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,248 @@
+                Set true to indicate if groundscatter should be plotted in grey
+                Default: False
+            zmin: int
+                Minimum value for colouring
```
``` suggestion
                The minimum parameter value for coloring
```

mts299 commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,248 @@
+                Minimum value for colouring
+                Default: {'p_l': [0], 'v': [-200], 'w_l': [0], 'elv': [0]}
+            zmax: int
+                Maximum value for colouring
```
``` suggestion
                The maximum parameter value for  coloring
```

mts299 commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,248 @@
+            zmax: int
+                Maximum value for colouring
+                Default: {'p_l': [50], 'v': [200], 'w_l': [250], 'elv': [50]}
+            colorbar: int
```
``` suggestion
        colorbar: matplotlib.pyplot.colorbar
```
Why is this an int?

mts299 commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,248 @@
+                Maximum value for colouring
+                Default: {'p_l': [50], 'v': [200], 'w_l': [250], 'elv': [50]}
+            colorbar: int
+                Draw a colourbar if set to an integer
```
```suggestion
                Draw a colourbar if None
```

mts299 commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,248 @@
+                the label that appears next to the colour bar
+                Default: ''
+        """
+        my_path = os.path.abspath(os.path.dirname(__file__))
```
```suggestion
        # TODO: put a function here to get the grid of the radar's FOV information 
        my_path = os.path.abspath(os.path.dirname(__file__))
```

mts299 commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,248 @@
+        # Setup scans for easy locating
+        scan_mark = [sub['scan'] for sub in dmap_data]
+        no_scans = 0
+        # TODO: my PEP8 linter says this variable is not used?
```
Is the TODO correct?

billetd replied:
Removed that line, I think it was accidentally left over from development

mts299 commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,248 @@
+        scan_mark = [sub['scan'] for sub in dmap_data]
+        no_scans = 0
+        # TODO: my PEP8 linter says this variable is not used?
+        scan_no = 0
```
``` suggestion 
```

mts299 commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,248 @@
+        # TODO: my PEP8 linter says this variable is not used?
+        scan_no = 0
+        beam_scan = np.zeros((len(dmap_data)))
+        for beam in range(len(dmap_data)):
```
Add a comment on what you are doing here please :)

billetd replied:
Resolved

mts299 commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,248 @@
+                scan[slist, iterat] = dmap_data[i.astype(int)][parameter]
+                grndsct[slist, iterat] = dmap_data[i.astype(int)]['gflg']
+                iterat += 1
+            except KeyError:
```
Why is this `KeyError` here? Please leave a comment :)

billetd replied:
Er... I was mostly copying the template from your code :P I don't actually know what it does, if it does something could you explain?

mts299 replied:
I think I had this for 'slist' haha well I guess I need to update that too

mts299 commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,248 @@
+            zmax = defaultzminmax[parameter][1]
+
+        # Setup plot
+        # TODO: eek! remeber there is an option to pass in a variable called ax
```
Did you see this TODO?

billetd replied:
Which one?

Not sure what to do in place of how cmap is implemented, as I'm very limited in Python knowledge right now

mts299 commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,248 @@
+                theta = [thetas[gates, beams], thetas[gates + 1, beams],
+                         thetas[gates + 1, beams + 1],
+                         thetas[gates, beams + 1]]
+                # TODO: try to avoid single letter names or at
```
review this TODO please :)

billetd replied:
I would say it's common to use `r` and `theta` as polar plot variables. Added comments

billetd commented:
@kherli & @jpreistad, the problem with the location of the FOV files has been fixed. After banging my head against a wall for a few hours, I just realised the MANIFEST file needed updating

billetd commented:
An update for everyone!

I've modified fan.py such that it no longer needs you to tell it which hemisphere you want to plot.

@aburrell & @mts299, documentation has been written and added (see: [plot_radar_fov.md](https://github.com/SuperDARN/pydarn/blob/plotting/fan/docs/user/plot_radar_fov.md)).

mts299 commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,248 @@
+    def __str__(self):
+        return "This class is static class that provides"\
+                " the following methods: \n"\
+                "   - plot_fan()\n"\
```
@aburrell @asreimer @billetd thoughts? 

When I see `radar_fov` I just think the fov plot, not the data.

mts299 commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,248 @@
+                "   - return_beam_pos()\n"
+
+    @classmethod
+    def plot_radar_fov(cls, dmap_data: List[dict], scan_index: int = 0,
```
whoops, wrong comment box.

@aburrell @asreimer @billetd thoughts? 

When I see `radar_fov` I just think the fov plot, not the data.

billetd replied:
I'm open to changing this. A seperate method for just the FOV would also be pretty simple to add.

carleyjmartin replied:
I think I agree, it would be SUPER useful to me going forward if I could use a method in pydarn to grab just the FOV easily.

mts299 replied:
I think FOV plotter would be good to just plot a radar region. Not in this PR though haha

aburrell replied:
Yes, put an issue in for that!  Also, it's a thing I use all the time.  I agree with @mts299 about the name, because I now realise in that in the back of my mind this is the type of plot I was visualising for this PR.

billetd replied:
Updated the name to `plot_fan`. Dedicated FOV plotter will come in a seperate PR.

carleyjmartin commented on docs/user/plot_radar_fov.md:
````diff
@@ -0,0 +1,70 @@
+
+```
+Here I plotted the 27th scan in the file, and because I didn't tell it which parameter to use, it has defaulted to line-of-sight velocity:
+![](../imgs/fan_1.png)
````
Fan1 and fan2 examples have an overlap with the colourbar and the plot. Maybe set figsize to a standard value, or give the option for user to set a figsize (I think this was done for the range time and summary plots?).

carleyjmartin commented on docs/user/plot_radar_fov.md:
```diff
@@ -0,0 +1,70 @@
+Here I plotted the 27th scan in the file, and because I didn't tell it which parameter to use, it has defaulted to line-of-sight velocity:
+![](../imgs/fan_1.png)
+
+Default plots also do not show groundscatter as grey. Set it to true to colour groundscatter this way:
```
So the default is not to grey groundscatter? 
The rest of the plots have groundscatter greyed as default, it might be nice to have them all standardised?
I guess groundscatter=True will also work if its a bool field?

carleyjmartin commented on docs/user/plot_radar_fov.md:
````diff
@@ -0,0 +1,70 @@
+
+Default plots also do not show groundscatter as grey. Set it to true to colour groundscatter this way:
+```python
+a = pydarn.fan.plot_radar_fov(fitacf_data, scan_index=27, groundscatter=1)
````
I know I mentioned this earlier, and Dan said it would be in a future release. But to document that discussion start/end time rather than scan_index might be more intuitive. 
(to grab the nearest scan to time or something, but also plotting partial scans would be cool, with something Devin was talking about doing)

carleyjmartin commented on docs/user/plot_radar_fov.md:
````diff
@@ -0,0 +1,70 @@
+plt.show()
+
+```
+![](../imgs/fan_2.png)
````
It also might be nice to have an auto title, with radar and time?
Also mentioned in person would be to have the radar name next to the dot for radar position - and continents underneath, but as you said that for a future release.

carleyjmartin commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,243 @@
+    @classmethod
+    def plot_radar_fov(cls, dmap_data: List[dict], scan_index: int = 0,
+                       all_gates: bool = False, parameter: str = 'v',
+                       lowlat: int = 50, cmap: str = None, groundscatter: bool = False,
```
See earlier comment about groundscatter: bool =True?

aburrell commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,243 @@
+                       lowlat: int = 50, cmap: str = None, groundscatter: bool = False,
+                       zmin: int = None, zmax: int = None,
+                       colorbar: plt.colorbar = None,
+                       colorbar_label: str = ''):
```
Another aspect that seems to be missing is beam range.

aburrell commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,243 @@
+                "   - return_beam_pos()\n"
+
+    @classmethod
+    def plot_radar_fov(cls, dmap_data: List[dict], scan_index: int = 0,
```
```suggestion
    def plot_radar_fov(cls, dmap_data: List[dict], ax=None, scan_index: int = 0,
```

Not sure how to do type suggestions with NoneType/axes class input.

aburrell commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,243 @@
+        # Setup plot
+        # TODO: eek! remeber there is an option to pass in a variable called ax
+        # This may screw up references
+        ax = plt.axes(polar=True)
```
```suggestion
        if ax is None:
            ax = plt.axes(polar=True)
```

aburrell commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,243 @@
+        else:
+            ax.set_ylim(-90, -lowlat)
+            ax.set_yticks(np.arange(-lowlat, -90, -10))
+        ax.set_xticklabels(['00', '', '06', '', '12', '', '18', ''])
```
Recommend using the style '18:00' or '1800'.  What do others think?

billetd replied:
I've always been used to it the way it's written, but open to change if people would rather the hourly format.

aburrell commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,243 @@
+        else:
+            rangelimit = 75
+
+        # Plotting! WOO!
```
Recommend making this comment a bit more informative for ESL programmers

billetd replied:
Fixed

aburrell commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,243 @@
+            colorbar_label: str
+                the label that appears next to the colour bar
+                Default: ''
+        """
```
Missing `Returns`

billetd commented:
Big update this morning with a bunch of changes based on reviews. Rest of changes based on reviews will come soon:

2020/06/22
- Renamed `plot_radar_fov` method to `plot_fan`. All references to it should also be fixed.
- Thanks to a suggested change by @aburrell, can now easily plot multiple fans on same plot.
- `all_gates` keyword has been removed, replaced with `ranges`. An inputted two element list of the lower and upper ranges to plot (default is [0,75]).
- Behaviour of colorbar keyword has changed. Now simply "True" or "False" to display colorbar.
- Added the `boundary` keyword. "True" by default, but when set to "False", doesn't plot the outline of the fan plot.
- By setting four variables to take the return from `plot_fan`, you will now get numpy arrays of the AACGM latitude and longitude, data, and groundscatter information.
- Updates docs with new features and changes to existing things.

Here's an example showing off multiple fans, custom ranges, and no boundary:

<img width="596" alt="fan_3" src="https://user-images.githubusercontent.com/26160732/85321697-5307f500-b482-11ea-89cd-046241fe1ea1.png">

billetd commented on docs/user/plot_radar_fov.md:
````diff
@@ -0,0 +1,70 @@
+```
+With the FITACF data loaded as a list of dictionaries (`fitacf_data` variable in above example), you may now call the `plot_radar_fov` method. Make sure you tell it which scan (numbered from first recorded scan in file) you want using `scan_index`:
+```python
+a = pydarn.fan.plot_radar_fov(fitacf_data, scan_index=27)
````
Fixed

billetd commented:
Just updated this with a fix for radars which are scanning backwards, which originally wasn't handled correctly.

Where we're at with this: 
- Scan locating is being pulled out of this PR, and is now in #69.
- Once that has been merged to develop, I'll update this PR to make use of it.
- There is also the scope to pull out the "scan making" inside of this code, before merging to develop. I.e., making a new PR which handles generating the arrays of data used for plotting. This seems like a good fit for the new `scan.py' file.

aburrell commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,233 @@
+			if colorbar_label != '':
+				cb.set_label(colorbar_label)
+
+		return beam_corners_aacgm_lats, beam_corners_aacgm_lons, scan, grndsct
```
Recommend also returning dtime

aburrell replied:
```suggestion
		return dtime, beam_corners_aacgm_lats, beam_corners_aacgm_lons, scan, grndsct
```

aburrell reviewed (commented):
Also, the .DS_Store file should be removed and this type of file should be added to the .gitignore.

aburrell commented:
I needed to make some fan plots, so left more comments.  All I needed was the boundaries, so the format was awkward.  However, by loading data and removing all the plotted data I was able to get what I needed.  Fortunately, I didn't need to highlight beams in this case.

![mcm_polar_cap_example](https://user-images.githubusercontent.com/7045886/93909297-a120d480-fccd-11ea-8ecb-080a25dc5d3f.png)

billetd commented:
> I needed to make some fan plots, so left more comments. All I needed was the boundaries, so the format was awkward. However, by loading data and removing all the plotted data I was able to get what I needed. Fortunately, I didn't need to highlight beams in this case.

You can now get the boundaries from a seperate function now (see PR #73), so that should make things much easier for most people now.

Thanks for the additional comments, honestly I never even thought that more people might not want polar plots (or rather, would want a polar plot but zoomed in on the radar FOV in question). I'll set it up so the axis object can be passed it, but leave non polar capabilities until after this is merged in. It should be striaghtforward though once the axis stuff is done.

billetd commented:
Updates as of today:

- Merged in from develop to include pyDARNio and beam position finding changes.
- Now accepts axis object as a keyword instead of overwriting. If `ax` keyword is left blank, it will make a new polar plot axis object.
- datetime object now also returned.

In a future update, a change will be to allow any time of map project as an axis. But for now, it is limited to only polar projections.

I believe this could be ready to merge now assuming nobody runs into more problems. @aburrell, @mts299 do this changes look okay?

mts299 commented on docs/user/plot_fan.md:
```diff
@@ -0,0 +1,86 @@
+# Fan plots
+---
+
+Fan plots are a way to visualise data from the entire scan of SuperDARN radar. It is when the data in all beams and ranges for a given parameter (such as line-of-sight velocity, backscatter power, etc) and a particular scan is projected onto a magnetic local time/magnetic latitude plot in [AACGMv2 coordinates](http://superdarn.thayer.dartmouth.edu/aacgm.html).
```
``` suggestion
Fan plots are a way to visualise data from the entire scan of a SuperDARN radar. Fan plots, plot all the data in all beams and ranges for a given parameter (such as line-of-sight velocity, backscatter power, etc) and a particular scan.
The data is projected onto a magnetic local time/magnetic latitude plot in [AACGMv2 coordinates](http://superdarn.thayer.dartmouth.edu/aacgm.html).
```

mts299 commented on docs/user/plot_fan.md:
```diff
@@ -0,0 +1,86 @@
+
+Fan plots are a way to visualise data from the entire scan of SuperDARN radar. It is when the data in all beams and ranges for a given parameter (such as line-of-sight velocity, backscatter power, etc) and a particular scan is projected onto a magnetic local time/magnetic latitude plot in [AACGMv2 coordinates](http://superdarn.thayer.dartmouth.edu/aacgm.html).
+
+Currently, fan plots in pyDARN get the geographic positions of a radars range gates automatically by reading in pre-generated files (found in the `/pydarn/radar_fov_files` folder), and then [converts](https://pypi.org/project/aacgmv2/) them to AACGMv2 coordinates. This will change in the future, such that pyDARN will work out the positions itself. That will bring suport for not standard range/bea ...
```
``` suggestion
Currently, fan plots in pyDARN get the geographic positions of a radar's range gates by reading in pre-generated files (found in the `/pydarn/radar_fov_files` folder), and then [converts](https://pypi.org/project/aacgmv2/) them to AACGMv2 coordinates. In the future, pyDARN will generate the geographical position, that will bring support for not standard range/beam layouts.
```

mts299 commented on docs/user/plot_fan.md:
````diff
@@ -0,0 +1,86 @@
+plt.show()
+
+```
+Here I plotted the 27th scan in the file, and because I didn't tell it which parameter to use, it has defaulted to line-of-sight velocity:
````
``` suggestion
In this example, the 27th scan was plotted with the defaulted parameter being line-of-sight velocity:
```

mts299 commented on docs/user/plot_fan.md:
````diff
@@ -0,0 +1,86 @@
+fitacf_data = SDarn_read.read_fitacf()
+
+```
+With the FITACF data loaded as a list of dictionaries (`fitacf_data` variable in above example), you may now call the `plot_fan` method. Make sure you tell it which scan (numbered from first recorded scan in file) you want using `scan_index`:
````
Does your code increment with 0 or 1? might need to mention what is the first scan number.

mts299 commented on docs/user/plot_fan.md:
```diff
@@ -0,0 +1,86 @@
+Here I plotted the 27th scan in the file, and because I didn't tell it which parameter to use, it has defaulted to line-of-sight velocity:
+![](../imgs/fan_1.png)
+
+Default plots also do not show groundscatter as grey. Set it to true to colour groundscatter this way:
```
Just checking since this happened to me, should velocity be the default with groundscatter? Or do Fan plots get used more for other parameters. 

I really don't know to be honest :p

carleyjmartin replied:
I think I brought this up in a review earlier, not sure what the outcome of that discussion was but I always thought ground scatter greyed out was default for velocity at least. I.E. the majority of people would want to know what is ground scatter when they're looking at the velocity. But then in the other three parameters not so much, but also the other three are much less used.

I think the range-time plots are default grey for velocity only? Depends if we want that coding consistency too?

It's a hard call to make really. For reference, our real time fan plots have ground scatter greyed out for velocity only with no option to un-grey it and we've had no complaints thus far.

mts299 commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,216 @@
+from pydarn import PyDARNColormaps, build_scan, radar_fov
+
+class Fan():
+    # TODO: Add class documentation
```
Still have this TODO to finish

mts299 commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,216 @@
+            zmin: int
+                The minimum parameter value for coloring
+                Default: {'p_l': [0], 'v': [-200], 'w_l': [0], 'elv': [0]}
+            zmax: int
+                The maximum parameter value for  coloring
+                Default: {'p_l': [50], 'v': [200], 'w_l': [250], 'elv': [50]}
+        	colorbar: bool
```
```suggestion
            zmax: int
                The maximum parameter value for  coloring
                Default: {'p_l': [50], 'v': [200], 'w_l': [250], 'elv': [50]}
            colorbar: bool
```

mts299 commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,216 @@
+			zmax = defaultzminmax[parameter][1]
+
+        # Setup plot
+        # TODO: eek! remeber there is an option to pass in a variable called ax
```
``` suggestion
```

mts299 commented on pydarn/plotting/fan.py:
```diff
@@ -0,0 +1,216 @@
+        # This may screw up references
+		if ax is None:
+			ax = plt.axes(polar=True)
+			if beam_corners_aacgm_lats[0,0] > 0:
```
Does this need to be back an ident because if `ax` is passed it won't work?

mts299 reviewed (commented):
Few suggestions and a couple of TODOs and maybe 1 potential issue.

billetd commented:
@mts299, @aburrell, would be good to give this one last check over/test before merging?

mts299 commented:
@billetd I cannot run your code seems like you have tabs and spaces in your files:
```bash
non-default coefficient files may be specified by running aacgmv2.wrapper.set_coeff_path before any other functions
Traceback (most recent call last):
  File "plot_fan.py", line 1, in <module>
    import pydarn
  File "/usr/lib/python3.6/site-packages/pydarn/__init__.py", line 42, in <module>
    from .plotting.fan import Fan
  File "/usr/lib/python3.6/site-packages/pydarn/plotting/fan.py", line 29
    def __str__(self):
                     ^
```
Would suggest using flake8 and trying to run it on maxwell?

billetd commented:
@mts299 should be fixed now

mts299 commented:
@billetd 

Sooo tried to run it and new error came up:

``` bash
python3 plot_fan.py 
resetting environment variable IGRF_COEFFS in python script
resetting environment variable AACGM_v2_DAT_PREFIX in python script
non-default coefficient files may be specified by running aacgmv2.wrapper.set_coeff_path before any other functions
Traceback (most recent call last):
  File "plot_fan.py", line 5, in <module>
    pydarn.Fan.plot_fan(fitacf_data)
  File "/usr/lib/python3.6/site-packages/pydarn/plotting/fan.py", line 228, in plot_fan
    cb = ax.figure.colorbar(mappable, ax=ax, extend='both', ticks=ticks)
  File "/usr/lib64/python3.6/site-packages/matplotlib/figure.py", line 2129, in colorbar
    cb = cbar.colorbar_factory(cax, mappable, **cb_kw)
  File "/usr/lib64/python3.6/site-packages/matplotlib/colorbar.py", line 1556, in colorbar_factory
    cb = Colorbar(cax, mappable, **kwargs)
  File "/usr/lib64/python3.6/site-packages/matplotlib/colorbar.py", line 1062, in __init__
    mappable.autoscale_None()
  File "/usr/lib64/python3.6/site-packages/matplotlib/cm.py", line 361, in autoscale_None
    raise TypeError('You must first set_array for mappable')
TypeError: You must first set_array for mappable
```

Code:
``` python
import pydarn

data = "./20200923.1001.00.rkn.fitacf"
fitacf_data = pydarn.SuperDARNRead(data).read_fitacf()
pydarn.Fan.plot_fan(fitacf_data)
```

Let me know if you have time to look into it. If not I can probably look at it tomorrow morning

carleyjmartin commented:
Worked for me now:
```python
import matplotlib.pyplot as plt
import pydarn

#Read in fitACF file
file = "/Users/carley/Documents/data/20131031.2201.00.sas.fitacf"
SDarn_read = pydarn.SuperDARNRead(file)
fitacf_data = SDarn_read.read_fitacf()

#Make the fan plot and show
fanplot = pydarn.Fan.plot_fan(fitacf_data, scan_index=1, groundscatter=1, colorbar='True', parameter='p_l', colorbar_label='Power [dB]')
plt.show()
```
With user warning:
```
173: UserWarning: FixedFormatter should only be used together with FixedLocator
```

Got this plot:
<img width="530" alt="Screen Shot 2021-01-14 at 11 06 27 AM" src="https://user-images.githubusercontent.com/60905856/104623956-91504600-5658-11eb-8f1f-06ab72cd105b.png">

mts299 commented:
Ran @carleyjmartin code and got:
``` bash
python3 plot_fan.py 
resetting environment variable IGRF_COEFFS in python script
resetting environment variable AACGM_v2_DAT_PREFIX in python script
non-default coefficient files may be specified by running aacgmv2.wrapper.set_coeff_path before any other functions
posx and posy should be finite values
posx and posy should be finite values
posx and posy should be finite values
posx and posy should be finite values
posx and posy should be finite values
posx and posy should be finite values
posx and posy should be finite values
posx and posy should be finite values
posx and posy should be finite values
posx and posy should be finite values
posx and posy should be finite values
posx and posy should be finite values
posx and posy should be finite values
posx and posy should be finite values
posx and posy should be finite values
posx and posy should be finite values
```

With this file: ./20200923.1001.00.rkn.fitacf

Wondering if my previous error was due to a parameter not being set?

billetd commented:
I ran @carleyjmartin's code on the same file and it worked fine, no errors for me.

@mts299 could you send me your file? Also try Carley's file.

mts299 commented:
@billetd I got it to work seems like an issue on me installing the old one

mts299 commented:
Shall Merge it now?
